New approach for enabling unit testing using virtual methods

gled4er opened the pull request:
Hello @cgillum 

I am opening this new pull request based on your feedback provided in previous pull request - https://github.com/Azure/azure-functions-durable-extension/pull/116

This pull request is work in progress for implementing issue - https://github.com/Azure/azure-functions-durable-extension/issues/64

I changed the public methods to virtual and modified the access levels for few constructors and few classes.

**I am very interested in your opinion especially for these changes**:
- enum FunctionType from **internal** to **public**
- class EndToEndTraceHelper from **internal** to **public**
- removing **sealed** from class **DurableOrchestrationContext**

My idea with the changes in this pull request is to enable developers to implement the following mocks:

```csharp

 public class DurableOrchestrationClientMock : DurableOrchestrationClient
    {
        public DurableOrchestrationClientMock(IOrchestrationServiceClient serviceClient, DurableTaskExtension config,
            OrchestrationClientAttribute attribute, EndToEndTraceHelper traceHelper) : base(serviceClient, config,
            attribute, traceHelper)
        {

        }

        public override async Task<DurableOrchestrationStatus> GetStatusAsync(string instanceId)
        {
            await Task.Delay(5);
            return new DurableOrchestrationStatus();
        }
    }


    public class DurableActivityContextMock : DurableActivityContext
    {
        public DurableActivityContextMock(string instanceId, string serializedInput) : base(instanceId, serializedInput)
        {
           
        }

        public override string InstanceId => "123";

        public override string GetRawInput()
        {
            return "Raw Input";
        }
    }

    public class DurableOrchestrationContextMock : DurableOrchestrationContext
    {
        public DurableOrchestrationContextMock(
            DurableTaskExtension config,
            string functionName,
            string functionVersion) : base(config, functionName, functionVersion)
        {
          
        }

        public override DateTime CurrentUtcDateTime => new DateTime();

        public override Task CallActivityAsync(string functionName, object input)
        {
            return Task.Run(() => "Hello World!");
        }
    }

```

Thank you!

cgillum commented on src/WebJobs.Extensions.DurableTask/DurableOrchestrationContext.cs:
```diff
@@ -105,7 +105,7 @@ internal void AssignToCurrentThread()
         /// The raw JSON-formatted orchestrator function input.
         /// </returns>
-        public string GetRawInput()
+        public virtual string GetRawInput()
```
Let's avoid exposing `GetRawInput`. I don't think it's a common scenario.

cgillum commented on src/WebJobs.Extensions.DurableTask/DurableOrchestrationContext.cs:
```diff
@@ -117,7 +117,7 @@ public string GetRawInput()
         /// The parsed <c>JToken</c> representation of the orchestrator function input.
         /// </returns>
-        public JToken GetInputAsJson()
+        public virtual JToken GetInputAsJson()
```
Let's avoid exposing `GetInputAsJson`. I don't think it is a common scenario.

cgillum commented on src/WebJobs.Extensions.DurableTask/DurableOrchestrationContext.cs:
```diff
@@ -80,7 +80,7 @@ public sealed class DurableOrchestrationContext
         /// <c>true</c> if the orchestrator function is currently being replayed; otherwise <c>false</c>.
         /// </value>
-        public bool IsReplaying => this.innerContext.IsReplaying;
+        public virtual bool IsReplaying => this.innerContext.IsReplaying;
```
Let's move `IsReplaying` into the abstract base class as a virtual property with a default value of `false`.

cgillum commented on src/WebJobs.Extensions.DurableTask/DurableOrchestrationClient.cs:
```diff
@@ -57,7 +57,7 @@ public HttpResponseMessage CreateCheckStatusResponse(HttpRequestMessage request,
         /// The specified function does not exist, is disabled, or is not an orchestrator function.
         /// </exception>
-        public Task<string> StartNewAsync(string orchestratorFunctionName, object input)
+        public virtual Task<string> StartNewAsync(string orchestratorFunctionName, object input)
```
Since this overload just calls the other `StartNewAsync` overload, let's move it into the base class and make it virtual. The more complex overload can be made abstract.

cgillum commented on src/WebJobs.Extensions.DurableTask/DurableOrchestrationClient.cs:
```diff
@@ -99,7 +99,7 @@ public Task<string> StartNewAsync(string orchestratorFunctionName, object input)
         /// <returns>A task that completes when the event notification message has been enqueued.</returns>
         [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1030:UseEventsWhereAppropriate")]
-        public async Task RaiseEventAsync(string instanceId, string eventName, object eventData)
+        public virtual async Task RaiseEventAsync(string instanceId, string eventName, object eventData)
```
Let's keep the implementation of `RaiseEventAsync` here but create an abstract method in a base class.

cgillum commented on src/WebJobs.Extensions.DurableTask/DurableOrchestrationClient.cs:
```diff
@@ -131,7 +131,7 @@ public Task<string> StartNewAsync(string orchestratorFunctionName, object input)
         /// <param name="reason">The reason for terminating the orchestration instance.</param>
         /// <returns>A task that completes when the terminate message is enqueued.</returns>
-        public async Task TerminateAsync(string instanceId, string reason)
+        public virtual async Task TerminateAsync(string instanceId, string reason)
```
Let's keep the implementation of `TerminateAsync` here but create a corresponding abstract method in the abstract base class.

cgillum commented on src/WebJobs.Extensions.DurableTask/DurableOrchestrationClient.cs:
```diff
@@ -149,7 +149,7 @@ public Task<string> StartNewAsync(string orchestratorFunctionName, object input)
         /// <param name="instanceId">The ID of the orchestration instance to query.</param>
         /// <returns>Returns a task which completes when the status has been fetched.</returns>
-        public async Task<DurableOrchestrationStatus> GetStatusAsync(string instanceId)
+        public virtual async Task<DurableOrchestrationStatus> GetStatusAsync(string instanceId)
```
Let's leave the implementation of `GetStatusAsync` here but create a corresponding abstract method in the abstract base class.

gled4er commented:
Hello @cgillum ,

Thank you for the detailed feedback.

I will work on the comments you gave me.

Thank you!

gled4er commented:
Hello @cgillum,

I followed your recommendations and I added two new base abstract classes. I wanted to clarify how you prefer to manage the comments for the methods that are included in both the abstract class and the derived class. Currently I have them in both places but I think in order to minimize the risk of going out of sync we should have them in one class only - for example only in the abstract class. What do you think? 

Please let me know if you see other issues that need to be fixed.

Thank you!

cgillum commented:
Excellent, thank you! For comments, I think the best way to minimize duplication between the parent and child class is to use `/// <inheritdoc />`.  For example:

```csharp
/// <inheritdoc />
public override HttpResponseMessage CreateCheckStatusResponse(HttpRequestMessage request, string instanceId)
{
    return this.config.CreateCheckStatusResponse(request, instanceId, this.attribute);
}
```

The API documentation builder will know to get the documentation from the base class.

gled4er commented:
Great! I will do that. Thank you for the quick response.

cgillum commented:
I quickly reviewed these changes and I think they look great. Can you add a few tests that demonstrate how they can be used?

gled4er commented:
Great idea! 

Sure, I will work on that next.

Thank you!

gled4er commented:
Hello @cgillum ,

I added a test project and 2 very simple tests. I mocked `DurableOrchestrationContext` for now. I wanted to hear your opinion about the Testing project. Now I am using the default Visual Studio Test Project but in the main project you use xUnit based one. What type of project do you prefer to use for the sample tests - class library or console app that is modified to support xUnit or another option? 

The other question is about the approach I am using. I changed the type of the parameter we use in the sample from `DurableOrchestrationContext` to `DurableOrchestrationContextBase` so I can mock an object and pass it to the function. Let me know if you have another idea for illustrating the use better:

```csharp
[FunctionName("E1_HelloSequence")]
        public static async Task<List<string>> Run(
            [OrchestrationTrigger] DurableOrchestrationContextBase context)
```

Thank you!

cgillum commented on src/WebJobs.Extensions.DurableTask/DurableOrchestrationClientBase.cs:
```diff
@@ -0,0 +1,81 @@
+using System.Net.Http;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.WebJobs.Extensions.DurableTask
```
>Microsoft.Azure.WebJobs.Extensions.DurableTask [](start = 10, length = 46)

Since this is a public type, let's change the namespace to Microsoft.Azure.WebJobs. I caught this when looking at your change to the sample.

cgillum commented on src/WebJobs.Extensions.DurableTask/DurableOrchestrationContextBase.cs:
```diff
@@ -0,0 +1,364 @@
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.WebJobs.Extensions.DurableTask
```
>Microsoft.Azure.WebJobs.Extensions.DurableTask [](start = 10, length = 46)

Let's change this to Microsoft.Azure.WebJobs.

cgillum commented on samples/VSSample.Tests/VSSample.Tests.csproj:
```diff
@@ -0,0 +1,191 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<Project ToolsVersion="15.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
```
>Project [](start = 1, length = 7)

It looks like you're using the legacy VS project file format. Can you please use the new .NET Standard project system, like we do with the other projects?

cgillum commented on samples/VSSample.Tests/packages.config:
```diff
@@ -0,0 +1,68 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<packages>
```
>packages [](start = 1, length = 8)

This file should not be needed when you convert to the new .NET Standard project system.

cgillum commented on samples/VSSample.Tests/VSSample.Tests.csproj:
```diff
@@ -0,0 +1,191 @@
+    <AppDesignerFolder>Properties</AppDesignerFolder>
+    <RootNamespace>VSSample.Tests</RootNamespace>
+    <AssemblyName>VSSample.Tests</AssemblyName>
+    <TargetFrameworkVersion>v4.6.1</TargetFrameworkVersion>
```
>TargetFrameworkVersion [](start = 5, length = 22)

Let's target .NET Standard for the target framework version, like we do for the unit test project.

cgillum commented on samples/precompiled/HelloSequence.cs:
```diff
@@ -4,14 +4,15 @@
 using System.Threading.Tasks;
 using Microsoft.Azure.WebJobs;
+using Microsoft.Azure.WebJobs.Extensions.DurableTask;
 
```
This new using statement should not be necessary.

cgillum commented on samples/VSSample.Tests/Properties/AssemblyInfo.cs:
```diff
@@ -0,0 +1,20 @@
+using System.Reflection;
```
Let's not add AssemblyInfo.cs for this project. I don't believe it is normally included when using the new .NET Standard project system.

cgillum commented:
> using System;

New source files should include the following header:

// Copyright (c) .NET Foundation. All rights reserved.
// Licensed under the MIT License. See LICENSE in the project root for license information.

---
Refers to: src/WebJobs.Extensions.DurableTask/DurableOrchestrationClientBase.cs:1 in 20439bd. [](commit_id = 20439bd19a5fe1106698542cd72ea69678426ad6, deletion_comment = False)

cgillum commented:
> using System;

New source files should include the following header:

// Copyright (c) .NET Foundation. All rights reserved.
// Licensed under the MIT License. See LICENSE in the project root for license information.

---
Refers to: src/WebJobs.Extensions.DurableTask/DurableOrchestrationContextBase.cs:1 in 20439bd. [](commit_id = 20439bd19a5fe1106698542cd72ea69678426ad6, deletion_comment = False)

gled4er commented:
Hello Chris,

Thank you very much for the detailed feedback.

I will fix all issues you commented.

Thank you!


Kanio

Sent from my Windows 10 phone

________________________________
From: Chris Gillum <notifications@github.com>
Sent: Saturday, January 6, 2018 4:50:15 AM
To: Azure/azure-functions-durable-extension
Cc: Kanio Dimitrov; Author
Subject: Re: [Azure/azure-functions-durable-extension] New approach for enabling unit testing using virtual methods (#117)


@cgillum requested changes on this pull request.

I like your idea to create a test project under samples. Let's go with that, but using the new .NET Standard project system.

I don't have a strong opinion on xunit vs mstest. I will let you decide. I went with xunit because that is what the rest of the Azure Functions project is using.

Can you also include some new automated tests in the WebJobs.Extensions.DurableTask.Test project? We want to make sure we don't accidentally regress the work that you did in the future.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FAzure%2Fazure-functions-durable-extension%2Fpull%2F117%23pullrequestreview-87006118&data=02%7C01%7CKanio.Dimitrov%40microsoft.com%7C99c3456e8481424c975208d554758ae4%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636507786176695712&sdata=XsbiOAOtxCw7qUzqYZublWH3TvTRRKPNfOcz3KD85zU%3D&reserved=0>, or mute the thread<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAEiVQMM-JV4eNW1PEXdEmfEmJuQMIo0eks5tHnz3gaJpZM4RO02u&data=02%7C01%7CKanio.Dimitrov%40microsoft.com%7C99c3456e8481424c975208d554758ae4%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636507786176695712&sdata=Bv%2BuSwH%2BCTpKp10vn3pHvZ%2BBbHjlzLVzzJt1g2EoHfc%3D&reserved=0>.

cgillum commented on test/DurableOrchestrationClientBaseTests.cs:
```diff
@@ -0,0 +1,25 @@
+            var result = await durableOrchestrationClientBaseMock.Object.StartNewAsync(functionName, null);
+            result.Should().Be(instanceId);
+        }
+    }
```
Should we add tests for the other abstract methods on `DurableOrchestrationClientBase` as well?

gled4er replied:
Yes, I will look into that.

cgillum commented on test/DurableOrchestrationClientBaseTests.cs:
```diff
@@ -0,0 +1,25 @@
+﻿using System;
```
These new test files will need to copyright headers as well.

gled4er replied:
I will add it. Thank you for catching that.

gled4er commented:
Hello @cgillum,

Great idea to add StyleCop! 

Of course I will fix the warning. 

Thank you for the review remarks.

gled4er commented:
Hello @cgillum ,

I started to add tests for DurableOrchestrationClient and the first two were ok. However I needed to modify the access to two method to protected:
```csharp
protected internal virtual HttpResponseMessage CreateCheckStatusResponse( ...

protected internal virtual async Task<HttpResponseMessage> WaitForCompletionOrCreateCheckStatusResponseAsync( ...

```
Please let me know if this is acceptable.

I wanted to apply the same approach to the rest methods in the `DurableOrchestrationClient` class but they depend on ` private readonly TaskHubClient client;`  which is initialized in the constructor  ` this.client = new TaskHubClient(serviceClient);`  . 

Currently I am thinking how to mock the  ` TaskHubClient`  variable but due to its private access level and the sealed definition of  ` TaskHubClient` I think this will be difficult. 

Please let me know if you see better approach or workaround. I think it will be great if we can abstract the initialization of  `client` via wrapping class but this will require further changes in the code and I want to align the vision with you first.

Thank you!

cgillum commented on samples/VSSample.Tests/HttpSyncStartTests.cs:
```diff
@@ -0,0 +1,77 @@
+{
+    public class HttpSyncStartTests
+    {
+        private const string FuncitonName = "SampleFunction";
```
Typo. `FuncitonName` should be `FunctionName`

gled4er replied:
Thank you for catching that!

cgillum commented on src/WebJobs.Extensions.DurableTask/DurableTaskExtension.cs:
```diff
@@ -491,8 +491,9 @@ internal string GetIntputOutputTrace(string rawInputOutputData)
-        internal HttpResponseMessage CreateCheckStatusResponse(
+
+        // Get a response that will point to our webhook handler. 
+        protected internal virtual HttpResponseMessage CreateCheckStatusResponse(
```
Why was it necessary to make this `protected` and `virtual`? I can't think of a good reason to expose anything in  `DurableTaskExtension`.

gled4er replied:
This was done in order to override the behavior in the mock class. I will revert these changes based on your feedback. Thank you!

cgillum commented on src/WebJobs.Extensions.DurableTask/DurableTaskExtension.cs:
```diff
@@ -505,9 +506,9 @@ internal string GetIntputOutputTrace(string rawInputOutputData)
-        internal async Task<HttpResponseMessage> WaitForCompletionOrCreateCheckStatusResponseAsync(
+        // Get a response that will wait for response from the durable function for predefined period of time before 
+        // pointing to our webhook handler. 
+        protected internal virtual async Task<HttpResponseMessage> WaitForCompletionOrCreateCheckStatusResponseAsync(
```
Same comment as previous.

gled4er replied:
I will revert these changes based on your feedback. Thank you!

cgillum commented on test/DurableOrchestrationClientTests.cs:
```diff
@@ -0,0 +1,52 @@
+
+namespace WebJobs.Extensions.DurableTask.Tests
+{
+    public class DurableOrchestrationClientTests
```
I don't think it's necessary to add unit testing for `DurableOrchestrationClient`. I imagine we already have good coverage elsewhere, and I feel a little uncomfortable about mocking some of the internal implementation details, like `EndToEndTraceHelper` and `DurableTaskExtension`. Can we leave these changes out of the PR for now? The main goal is to enable developers to write mocks for the public surface area, and I think your other changes have already achieved and tested this.

gled4er replied:
Thank you for the detailed feedback here. I will revert these changes based on your feedback. Thank you!

gled4er commented:
Hello @cgillum ,

Thank you for the detailed feedback. 

I will revert the changes around `DurableOrchestrationClient`. 

I needed to confirm first with you before making the changes. 

When you mentioned in a previous comment - **"Should we add tests for the other abstract methods on `DurableOrchestrationClientBase` as well?"**,  what did you have in mind?  Initially I added unit tests only for the methods in the abstract class that were calling overloads to ensure that we have these changes tested. So it will be great if you give me few more hints how to test the pure abstract methods in the abstract classes inside the main project. 

Thank you very much!

cgillum commented:
I apologize, I think I misunderstood the intent of your tests. I understand now that you were specifically testing abstract methods which were calling other overloads. I was thinking we wanted to mock the pure abstract methods as well to ensure that any breaking changes to these abstract APIs would be caught by the tests. It may be a bit overkill, so we don't have to add any more tests than what you have.

If you can revert the latest changes around `DurableOrchestrationClient`, then I think I can sign off on this PR.

gled4er commented:
Hello @cgillum ,

Thank you for the quick reply!

I am already working on reverting the changes.

Thank you!

gled4er commented:
Hello @cgillum ,

I reverted the test changes around `DurableOrchestrationClient'. Style Cop warnings are also addressed.

May I ask you to review again and let me know what should be changed next?

Thank you!

gled4er commented:
Hello @cgillum ,

Thank you for the great feedback and guidance! 

I will be happy to help with the documentation. Let me know if you have some suggestions in terms of location in the existing documentation pages and the depth of the guidance. 

Maybe we can use the samples' code as foundation and provide explanations around it?

Thank you!

cgillum commented:
Actually, before we merge, can you verify that the `E1_HelloSequence` sample can still run when using `DurableOrchestrationContextBase` instead of `DurableOrchestrationContext`? I have a feeling it might not and that another small change may be necessary.

gled4er commented:
Hello @cgillum ,

It works well. I tested if both from `HttpSyncStart` and from `HttpSyncStart`. 

I am not able to trace it exactly at the moment but I think that `DurableOrchestrationClientBase starter` is assigned behind the scenes to instance of `DurableOrchestrationClient` so the code is actually acting as IoC container if that case. 

Let me know if you want me to look deeper into it.

Thank you!
